### PR TITLE
Tests for scripts in io/fits/scripts

### PR DESCRIPTION
**Pull request by mohanagr**
Implements test for script `fitsheader` as of now.

### Reference:
Issue #5799

### CODE REVIEW

**Comment by mohanagr** [comment]
Table formatting is yet to be tested. Working on that.

**Comment by MSeifert04** [comment]
I think some unrelated changes (commits) made it into this PR. Could you clean them up (remove them) so it only contains "new" changes?

**Comment by mohanagr** [comment]
Oh. My bad. Sorry.

File: astropy/io/fits/tests/test_fitsheader.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        sys.argv = self.sys_argv_orig
+
+    def test_noargs(self):
+        with pytest.raises(SystemExit) as e:
+            fitsheader.main()
+        assert e.value.code == 2
+
+    @pytest.mark.parametrize('test_file', [
+        ('test0.fits'),
+        pytest.mark.xfail(('random.fits'), reason = 'bad file')
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_fitsheader.py]: Why create a test just to `xfail` it? Or am I misunderstand this line?

File: astropy/io/fits/tests/test_fitsheader.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+        fitsheader.main()
+        out, err = capsys.readouterr()
+
+        assert expected in out
+    
+    # Different test because stderr will not be none in case of testing by existing keyword
+    # fitsheader searches for a keyword in all the HDU's
+    @pytest.mark.parametrize('test_kw', [
+        ('RANDOMKEY')
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_fitsheader.py]: This looks like only one test? It doesn't make sense to actually `parametrize` the test, right?

File: astropy/io/fits/tests/test_fitsheader.py (hunk 1 above)
Comment (mohanagr) [inline astropy/io/fits/tests/test_fitsheader.py]: There are two tests, one is supposed to pass i.e. `test0.fits` does exist and other `xfails`.

File: astropy/io/fits/tests/test_fitsheader.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+from ....tests.helper import pytest
+from ....utils.exceptions import AstropyDeprecationWarning
+from ....version import version
+ 
+# @pytest.fixture(scope=module , params=['./tests/test0.fits'])
+# def resource_testfile_setup(request):
+#     with open(request.params[0]) as test_fits:
+#         yield test_fits
+
+@pytest.fixture(scope='module')
```
Comment (MSeifert04) [inline astropy/io/fits/tests/test_fitsheader.py]: Is it actually necessary to declare this helper-function as fixture?

File: astropy/io/fits/tests/test_fitsheader.py (hunk 1 above)
Comment (MSeifert04) [inline astropy/io/fits/tests/test_fitsheader.py]: But a test that should fail shouldn't be `xfail` but test it with `pytest.raises`?

File: astropy/io/fits/tests/test_fitsheader.py (hunk 3 above)
Comment (mohanagr) [inline astropy/io/fits/tests/test_fitsheader.py]: Using fixture I wouldn't have to call the function everytime. (lesser executions because `module` scope? lots of method use it).

File: astropy/io/fits/tests/test_fitsheader.py (hunk 1 above)
Comment (bsipocz) [inline astropy/io/fits/tests/test_fitsheader.py]: I agree with @MSeifert04. ``xfail`` shouldn't be used for this case, if a failure is expected and normal behaviour then it should be caught with ``pytest.raises``.

File: astropy/io/fits/tests/test_fitsheader.py (hunk 3 above)
Comment (MSeifert04) [inline astropy/io/fits/tests/test_fitsheader.py]: So this is just a way so the regex doesn't need to be compiled multiple times?

Comment (mohanagr) [inline astropy/io/fits/tests/test_fitsheader.py]: Yes. module scope fixtures distribute the return value across functions which need it, don't they? Although I know it's almost no overhead. Should I remove it and replace by functions?

Comment (MSeifert04) [inline astropy/io/fits/tests/test_fitsheader.py]: Well they seem like a really complicated way (using regex) of testing wheater `str.startswith('WARNING')` or `'WARNING' in str`, right?

**Comment by MSeifert04** [comment]
It's my personal opinion but I wouldn't test any expected failures, like specifying the wrong file, non-existing keyword or non-existing extensions.

I think we basically need to make sure the script works and doesn't get accidentally broken. The failures (probably) come from the non-scripty part of the `io.fits` module and should be tested already there.

File: astropy/io/fits/tests/test_fitsheader.py (hunk 3 above)
Comment (mohanagr) [inline astropy/io/fits/tests/test_fitsheader.py]: I just thought RE might be of use later. I'll change it if required.

**Comment by mohanagr** [comment]
Doesn't script not emitting warning when it should come under _accidentally broken_ (i.e. bad user inputs and things)?

File: astropy/io/fits/tests/test_fitsheader.py (hunk 1 above)
Comment (mohanagr) [inline astropy/io/fits/tests/test_fitsheader.py]: @MSeifert04 @bsipocz All right. What are the use cases for `xfail` then?

Comment (MSeifert04) [inline astropy/io/fits/tests/test_fitsheader.py]: - `xfail` - should work but currently doesn't because of some unfixed bug (either in-package or upstream)
- `skip` - doesn't work if dependency is too old or if dependency is not installed
- `pytest.raises` - should fail, make sure it fails with the right exception (and exception message).

That's not official, that's just how I interpret these.

**Comment by MSeifert04** [comment]
@mohanagr Before you push anything else: I would be very interested to see what travis+coverage reports.

I really need to think a bit more about the scope of the PR and what to test with scripts. In the end some utility that can be used to test scripts would be really nice. Something like you basically did with your `setup` and `teardown` but more dynamic so it can be easily adjusted for other scripts.

Let's wait and see what the others say.

**Comment by mohanagr** [comment]
All right. I'd really like to contribute more to this issue and the testing part in general. Guess that'd help me slowly understand the codebase better, so that I can tackle other issues as well.

**Comment by mohanagr** [comment]
@MSeifert04 I got your point. The cases for invalid files, extensions etc. has been handled in the _non-script_  part of the script! I'll remove the `xfail` ones in the next commit once Travis finishes builds.

File: astropy/io/fits/tests/test_fitsheader.py (hunk 3 above)
Comment (pllim) [inline astropy/io/fits/tests/test_fitsheader.py]: I agree with @MSeifert04 . This seems unnecessarily complicated. Using `startswith()` is more readable.

File: astropy/io/fits/tests/test_fitsheader.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+        fitsheader.main()
+        out, err = capsys.readouterr()
+
+        assert error_regex.match(err) is None
+        assert warning_regex.match(err) is None
+
+    @pytest.mark.parametrize('test_kw,expected', [
+        ('BSCALE', 'BSCALE  =           1.000000E0 / REAL = TAPE*BSCALE + BZERO'),
+        pytest.mark.xfail(('LRFWAVE', 'LRFWAVE = 0.0 / linear ramp filter wavelength'), reason = 'bad format')
```
Comment (pllim) [inline astropy/io/fits/tests/test_fitsheader.py]: Putting `xfail` inside `parametrize` is hard to read. I prefer `xfail` to be its own thing.

File: astropy/io/fits/tests/test_fitsheader.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+@pytest.fixture(scope='module')
+def error_regex():
+    regex = re.compile('^ERROR.*')
+    return regex
+
+class TestFITSheader_script(FitsTestCase):
+
+    def setup_method(self,  method):
+        self.sys_argv_orig = sys.argv
```
Comment (pllim) [inline astropy/io/fits/tests/test_fitsheader.py]: Using `subprocess` to spawn the command line would be simpler but I wonder if that would work in Windows; do you know, @MSeifert04 ?

Comment (MSeifert04) [inline astropy/io/fits/tests/test_fitsheader.py]: There should be no problem with `subprocess`, I've used it several times (mostly to call `zlib`). However, is `coverage` able to deal with subprocesses?

Comment (pllim) [inline astropy/io/fits/tests/test_fitsheader.py]: Good point. Maybe this is not worth pursuing.

Comment (mohanagr) [inline astropy/io/fits/tests/test_fitsheader.py]: @pllim `subprocess` does work on windows AFAIK. Also I will remove the `xfail` things. I understand the usage better now. I will add the test for the table formatting and go on with `fitscheck`.

**Comment by saimn** [comment]
It is possible to avoid mocking `sys.argv` by having the `main` function take an optional `args` argument which is passed to `ArgumentParser.parse_args`. This is already possible with the fitsheader script: https://github.com/astropy/astropy/blob/master/astropy/io/fits/scripts/fitsheader.py#L284 (and easy to do for the other scripts).

About the way to test scripts, some options:
- pytest has to `TestDir.run` method to test itself it seems, in a subprocess or not. http://stackoverflow.com/a/13500346/4728722 gives an example.
- there is also https://pypi.python.org/pypi/pytest-console-scripts

Just some thoughts, I'm not sure it's useful to put too much work here.

**Comment by mohanagr** [comment]
@saimn @pllim In my opinion `monkeypatch` is the best thing to do. It allows one to change context for a test and then go back. (Like arhv, pwd, syspath etc. Etc.). Simple enough.

**Comment by mohanagr** [comment]
@saimn @pllim In my opinion `monkeypatch` is the best thing to do. It allows one to change context for a test and then go back. (Like arhv, pwd, syspath etc. Etc.). Simple enough.

**Comment by pllim** [comment]
For the sake of simplicity, I think the current call to `main()` should be okay...

**Comment by mohanagr** [comment]
@pllim Sorry I didn't read your comment after pushing. I have taken other comments into account.

- Removed `xfails`
- Fixed need to mock `sys.argv`
I'm deliberately including a few tests where the script should fail by making assertions on output (instead of raising an error myself and using `pytest.raises`) I'll remove those too if necessary.

**Comment by MSeifert04** [comment]
@mohanagr Just out of interest: Did you create tests based on (missing) coverage?

**Comment by mohanagr** [comment]
@MSeifert04 Not really. But coverage in this particular file?

**Comment by mohanagr** [comment]
Also can you suggest how to test `fitscheck`? I'll manually run that script and compare outputs?

**Comment by saimn** [comment]
> @saimn @pllim In my opinion monkeypatch is the best thing to do.

IMO it brings nothing here and only adds useless complexity, as you can use directly `fitsheader.main([...])` and it will work the same.

**Comment by mohanagr** [comment]
Can someone please tell me how to go ahead with #5874 ? Should I push the fix here?

**Comment by pllim** [comment]
Ideally, the bug fix would be a separate PR and should be merged first, and then this PR to be rebased on top of that once merged. They have to be separate because this PR would go into v2.0 and the bug fix into v1.3.2. Hope this clarifies the matter. I apologize for any inconvenience caused.

**Comment by mohanagr** [comment]
@pllim Yes. Thanks. I'll just file a separate PR too. I pushed here so that I could continue testing this. Or I'll to omit one test case. 
P.S. There is really no inconvenience really. Things for me to learn.

**Comment by pllim** [comment]
@mohanagr , the affected test can move to the new bug fix PR as well, if possible. This is because usually we ask the bug fix PR to add a test to demonstrate that the bug is indeed fix and to prevent it from creeping back in the future.

**Comment by saimn** [comment]
@mohanagr - Are you still interested in this ? If yes, to make progress, I would suggest to focus here on `test_fitsheader.py`, and to limit the number of things to test.

**Comment by mohanagr** [comment]
Yes, I'll continue this and try to fix the test scripts.

On May 15, 2017 3:43 PM, "Simon Conseil" <notifications@github.com> wrote:

@mohanagr <https://github.com/mohanagr> - Are you still interested in this
? If yes, to make progress, I would suggest to focus here on
test_fitsheader.py, and to limit the number of things to test.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/astropy/astropy/pull/5858#issuecomment-301434142>, or mute
the thread

File: astropy/io/fits/tests/test_fitsheader.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+import sys
+
+import numpy as np
+import re
+
+from . import FitsTestCase
+from ..hdu import PrimaryHDU
+from ..scripts import fitsheader
+from ....tests.helper import catch_warnings
+from ....tests.helper import pytest
```
Comment (saimn) [inline astropy/io/fits/tests/test_fitsheader.py]: Could you rebase on master, and import pytest directly (the bundle version was removed recently).

File: astropy/io/fits/tests/test_fitsheader.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+import sys
+
+import numpy as np
+import re
+
+from . import FitsTestCase
+from ..hdu import PrimaryHDU
```
Comment (saimn) [inline astropy/io/fits/tests/test_fitsheader.py]: Please remove the imports that are not used (`PrimaryHDU`, `AstropyDeprecationWarning`).

**Comment by saimn** [comment]
@mohanagr - Ok, great. So I think you should start with testing the normal usage of the script: choose a test file and simply test that the output is the expected one (with or without `-e`). 
With a test file whose header is not too long (I would choose `astropy/io/fits/tests/data/table.fits`) you can directly compare the output with the expected one. The goal is really to test the script, not the behavior of `fits.Header` which is already well tested. So there is no real need for all the pytest magic here (test one keyword with `-k` and remove the `parametrize` uses, remove `monkeypatch` as explained above), and keep the errors testing and edge cases for the end.

**Comment by pllim** [comment]
This was created during GSoC 2017 application period, which is now over. @mohanagr , if you are interested to finish this, feel free to re-open, but I am closing this for now due to inactivity.
